### ISSUE TITLE
Permit ngsw-bypass Header for Sentry Requests

### ISSUE THREAD

**Issue by Fwang36** [getsentry/sentry#47912]
### Problem Statement

Angular users require the `ngsw-bypass` header attached to their Sentry Requests to avoid caching Sentry requests by the Angular Service worker.  

Currently they are met with a CORS error of `...has been blocked by CORS policy: Request header field ngsw-bypass is not allowed by Access-Control-Allow-Headers in preflight response.`

[Angular Docs on this header](https://angular.io/guide/service-worker-devops#bypassing-the-service-worker)

### Solution Brainstorm

We should permit this header for Sentry Requests

### Product Area

Other

**Comment by muuvmuuv** [comment]
Hey, I contact the support.

This ticket here is related: #10481

The main issue is that requests from the service worker are blocked by Sentry, so the bypass parameter must be set. Unfortunately, Angular does not support whitelisting and the parameter "ngsw-bypass" can only be used in the request header or as a param.

While that would work, it is currently not possible (or no longer), to set this as a query parameter for Sentry.

Here is the relevant code part I think: https://github.com/getsentry/sentry-javascript/blob/15d9102bb18987e17ee602111fbf0435e8d9c6f5/packages/core/src/api.ts#L34-L49

**Comment by iker-barriocanal** [comment]
Hi @Fwang36 @muuvmuuv, I'm not sure if I understand the problem. Is Sentry blocking requests, or are they being cached by Angular and thus not reaching Sentry? Is the Angular SDK missing some instrumentation to allow specific request parameters?

**Comment by muuvmuuv** [comment]
Not cached I think but the request by Angular is blocked.

Here are a few cases:

- Works (patched param): https://entergon.app/preview/release/3.2.4/
- Does not due to request header blocking: https://entergon.app/preview/release/3.0.0/
- 504 Gateway due to NGSW: https://entergon.app/preview/release/2.6.1/

**Comment by iker-barriocanal** [comment]
I think this is something we should fix on the SDK side then, transferring to the JS SDK team.

**Comment by muuvmuuv** [comment]
This isnt completed?

**Comment by iker-barriocanal** [comment]
@stephanie-anderson not sure if the team got notified of this, could you take a look? Thank you!

**Comment by Lms24** [comment]
Hi @muuvmuuv this slipped through our (SDK team) radar, sorry about that! If I understand you correctly though, the problem is that the Sentry backend (where the SDKs send the requests that get blocked) doesn't accept this header. I'm assuming you're appending the `ngsw-bypass` header to outgoing Sentry requests? 

@iker-barriocanal I think we need to adjust Relay(?) CORS configuration to allow for this header. To me this sounds like this isn't something we can fix in the SDK.

**Comment by Lms24** [comment]
Oh I think I might have missed a crucial reply here. I have some questions:

> requests from the service worker are blocked by Sentry

why/how? What does the service worker do that makes Sentry block the requests?

> While that would work, it is currently not possible (or no longer), to set this as a query parameter for Sentry.

Okay, so does this mean, that part of the problem is that you cannot attach this http header OR the query parameter to the request made by the SDK? So what you're asking is that the SDK offers a possibility to attach this header for you? 

The thing I don't understand here is, if there's no way to attach the header, why did you get a CORS error saying that the header isn't allowed? 😅

**Comment by muuvmuuv** [comment]
@Lms24, that's OK, schedules are rough these days, I know that haha.

Yes, if I could attach a query param properly (see my patch), I don't get the error.

I cannot really tell you what Angular Service Worker does under the hood, thats something Angular need to tell. But I assume it could also be a browser limit to prevent SWs to attack servers?

**Comment by Lms24** [comment]
But I'm still wondering how you got the CORS error then?

**Comment by muuvmuuv** [comment]
From the [Angular Service Worker](https://angular.io/guide/service-worker-devops#service-worker-requests-when-the-server-cant-be-reached) docs:

> The service worker processes all requests unless the [service worker is explicitly bypassed](https://angular.io/guide/service-worker-devops#bypassing-the-service-worker). The service worker either returns a cached response or sends the request to the server, depending on the state and configuration of the cache. The service worker only caches responses to non-mutating requests, such as GET and HEAD.
If the service worker receives an error from the server or it doesn't receive a response, it returns an error status that indicates the result of the call. For example, if the service worker doesn't receive a response, it creates a [504 Gateway Timeout](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504) status to return. The 504 status in this example could be returned because the server is offline or the client is disconnected.

I cannot tell you exactly why Sentry response are "bad" and 504 is returned but that states its an issue with Service Worker or Sentry response. 

Are you able to have a look at your server log files and see if anything is wrong coming from "entergon.app/preview/release/2.6.1"?

**Comment by Lms24** [comment]
Okay, I did a little more digging and I _might_ have something but I need your confirmation on that @muuvmuuv 

> The service worker only caches responses to non-mutating requests, such as GET and HEAD.

This makes me think that the issue isn't the Service worker as all the requests made from the SDK to send Sentry events (errors, transactions, etc.) are `POST` requests. So the service worker shouldn't cache anything here and to my understanding there shouldn't be a need to bypass it in the first place. 

So I created a [minimal repro app](https://github.com/Lms24/gh-sentry-47912-angular-service-worker) with the following components:
* Barebones Angular 16 app
* Ran `ng add @angular/pwa`, which created all the NGSW config (I made no changes to it)
* Added and installed @sentry/angular-ivy as written in our docs (no custom changes, no `ngsw-bypass` header/param)
* Added a button to throw a test error. 

I [deployed it the app to vercel](https://gh-sentry-47912-angular-service-worker.vercel.app/) to test the service worker in action. 

At first, I managed to reproduce the problem: 

![image](https://github.com/getsentry/sentry/assets/8420481/6a3ce84a-7fdb-4561-a2f0-a7c3c56ce712)

Then I realized that I still had my ad blocker running. If I turn it off, all requests are sent to Sentry. If you go to the website and check the dev tools, you can try this for yourself. 

**So my question here is**: Is there any chance that this happens because of active ad blockers? The error is just a little obscure because the Service worker seems to change it to this generic 504 timeout error but it would make sense, considering:

> For example, if the service worker doesn't receive a response, it creates a [504 Gateway Timeout](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504) status to return.

If yes, you might want to set up a tunnel and use the `tunnel` option in the SDK. More on that can be found in our [Dealing with Ad-Blockers](https://docs.sentry.io/platforms/javascript/troubleshooting/#dealing-with-ad-blockers) guide.

If this is not the case, we need a way to properly reproduce the bug first. Please use the repro app I created as a base (feel  free to fork it or open PRs) to show us how the problematic behaviour can be reproduced.

**Comment by muuvmuuv** [comment]
Ok, this is wild... first: yes I did deactivate all ad blockers, this is a little obvious at first ^^ since Sentry is kinda a tracking service. 

BUT, I found something out. **ALL** browsers work, except Brave, so I tried a few things and I *disabled* ad blocking globally... and it works?! So _BRAVE_ is handling SW requests outside the domain, so deactivating ad blockers for that domain does simply not target service worker requests. Any since Angular SW just passes `POST` requests, its still a SW request :D wow!

The reason "ngsw-bypass" works, is just because in that case the domain does the request, where the deactivated ad-blocked does nothing.

Ok, to summarize, I think this can be closed. But I think you should add a note to the Troubleshooting section about Brave or in general browsers that handle sw/worker requests different. And tunneling would work, yes, because the ad blocker wouldn't hit.

**Comment by Lms24** [comment]
Wow, that's an interesting find 😅 

> in, so deactivating ad blockers for that domain does simply not target service worker requests. 

This sounds like a bug in Brave to me on first thought. I'd recommend submitting a [bug report](https://support.brave.com/hc/en-us/articles/360017916592-How-do-I-submit-a-bug-report-) with them.

> But I think you should add a note to the Troubleshooting section about Brave or in general browsers that handle sw/worker requests different

Hmm yes, thats worth a thought. I'll have a look at our docs around ad blockers. Maybe we include a warning about Brave there. In general though I think that tunneling is the best response here either way. 

> Ok, to summarize, I think this can be closed

I agree but thanks for sticking with us through this ride! Let me know if you have more questions.
